squash-the-creeps/rust/src/player.rs

```diff
@@ -0,0 +1,186 @@
+            //direction = direction.normalized()
+            direction = direction.normalized();
+
+            //TAKE THE PIVOT
+            let mut pivot = self.base_mut().get_node_as::<Node3D>("Pivot");
+            //$Pivot.basis = Basis.looking_at(direction) (GDScript)
+            pivot.set_basis(Basis::looking_at(-direction, Vector3::UP, true));
+            //$AnimationPlayer.speed_scale = 4
+            self.base()
+                .get_node_as::<AnimationPlayer>("AnimationPlayer")
+                .set_speed_scale(4.0);
+        } else {
+            //$AnimationPlayer.speed_scale = 1
+            self.base()
+                .get_node_as::<AnimationPlayer>("AnimationPlayer")
+                .set_speed_scale(1.0);
+        }
+
+        // Ground Velocity
+        //velocity.x = direction.x * speed
+        self.target_velocity.x = direction.x * self.speed;
+        //velocity.z = direction.z * speed
+        self.target_velocity.z = direction.z * self.speed;
+
+        //jumping
+        //if is_on_floor() and Input.is_action_just_pressed("jump"):
+        if self.base().is_on_floor() && input.is_action_just_pressed("jump") {
+            //velocity.y += jump_impulse
+            self.target_velocity.y = self.jump_impulse;
+        }
+
+        //We apply gravity every frame so the character always collides with the ground when moving.
+        //This is necessary for the is_on_floor() function to work as a body can always detect
+        //the floor, walls, etc. when a collision happens the same frame.
+
+        if !self.base().is_on_floor() {
+            //velocity.y -= fall_acceleration * delta
+            self.target_velocity.y -= self.fall_acceleration * _delta as f32;
+        }
+        //moving the Character
+        let velocity = self.target_velocity;
+        self.base_mut().set_velocity(velocity);
+        //move_and_slide()
+        self.base_mut().move_and_slide();
+
+        //Here, we check if we landed on top of a mob and if so, we kill it and bounce.
+        //With move_and_slide(), Godot makes the body move sometimes multiple times in a row to
+        //smooth out the character's motion. So we have to loop over all collisions that may have
+        //happened.
+        //If there are no "slides" this frame, the loop below won't run.
+        //for index in range(get_slide_collision_count()):
+        for index in 0..self.base().get_slide_collision_count() {
+            //var collision = get_slide_collision(index)
+            let collision = self.base_mut().get_slide_collision(index).unwrap();
+            //if collision.get_collider().is_in_group("mob"):
+            if let Some(collider) = collision.get_collider() {
+                if let Some(node) = collider.try_cast::<Node3D>().ok() {
+                    if node.is_in_group("mob") {
+                        //var mob = collision.get_collider()
+                        let mut mob = collision.get_collider().unwrap().cast::<Mob>();
```

Comment:
is there a case in which collider can be null :thinking:? I would use `expect(some reasoning)` instead of unwrap here :thinking: .